kani-compiler/src/kani_middle/reachability.rs

```diff
@@ -57,18 +59,35 @@ pub fn collect_reachable_items<'tcx>(
 }
 
 /// Collect all (top-level) items in the crate that matches the given predicate.
+/// An item can only be a root if they are: non-generic Fn / Static / GlobalASM
 pub fn filter_crate_items<F>(tcx: TyCtxt, predicate: F) -> Vec<MonoItem>
 where
-    F: FnMut(TyCtxt, DefId) -> bool,
+    F: Fn(TyCtxt, DefId) -> bool,
 {
-    let mut filter = predicate;
-    tcx.hir_crate_items(())
-        .items()
-        .filter_map(|hir_id| {
-            let def_id = hir_id.owner_id.def_id.to_def_id();
-            filter(tcx, def_id).then(|| MonoItem::Fn(Instance::mono(tcx, def_id)))
-        })
-        .collect()
+    let crate_items = tcx.hir_crate_items(());
+    // Filter regular items.
+    let root_items = crate_items.items().filter_map(|item| {
+        let def_id = item.owner_id.def_id.to_def_id();
+        if !is_generic(tcx, def_id) && predicate(tcx, def_id) {
+            to_mono_root(tcx, item, def_id)
+        } else {
+            None
+        }
+    });
+
+    // Filter items from implementation blocks.
+    let impl_items = crate_items.impl_items().filter_map(|impl_item| {
```

Comment:
Are `impl_items` not included in `items` above?

Reply:
That was my initial though, but I was wrong. For harnesses it didn't matter much, since we don't expect associated functions with them, but for other things they might be important.

FYI, for the little test I added in this PR, here are the breakdown of items and impl_items:
```
### ITEMS
def_id: DefId(0:1 ~ associated_fn[18c1]::{use#0}) -- kind: Use
def_id: DefId(0:2 ~ associated_fn[18c1]::std) -- kind: ExternCrate
def_id: DefId(0:3 ~ associated_fn[18c1]::Dummy) -- kind: Struct
def_id: DefId(0:5 ~ associated_fn[18c1]::{impl#0}) -- kind: Impl

### IMPL Items
def_id: DefId(0:6 ~ associated_fn[18c1]::{impl#0}::new) -- kind: AssocFn
```
Ps.: `associated_fn` is the name of the crate where they are defined. :)